subprojects/core/src/main/java/org/gradle/cache/internal/DefaultCrossBuildInMemoryCacheFactory.java

```diff
@@ -82,86 +86,84 @@ public <V> CrossBuildInMemoryCache<Class<?>, V> newClassMap() {
     }
 
     private abstract static class AbstractCrossBuildInMemoryCache<K, V> implements CrossBuildInMemoryCache<K, V>, BuildSessionLifecycleListener {
-        private final Object lock = new Object();
-        private final Map<K, V> valuesForThisSession = new HashMap<>();
+        private final ConcurrentHashMap<K, Lazy<V>> valuesForThisSession = new ConcurrentHashMap<>();
 
         @Override
         public void beforeComplete() {
-            synchronized (lock) {
-                retainValuesFromCurrentSession(valuesForThisSession.values());
-                valuesForThisSession.clear();
-            }
+            retainValuesFromCurrentSession(valuesForThisSession.values().stream().map(Lazy::get));
+            valuesForThisSession.clear();
         }
 
         @Override
         public void clear() {
-            synchronized (lock) {
-                valuesForThisSession.clear();
-                discardRetainedValues();
-            }
+            valuesForThisSession.clear();
+            discardRetainedValues();
         }
 
-        protected abstract void retainValuesFromCurrentSession(Collection<V> values);
+        protected abstract void retainValuesFromCurrentSession(Stream<V> values);
 
         protected abstract void discardRetainedValues();
 
+        /**
+         * Must be thread-safe.
+         */
         protected abstract void retainValue(K key, V v);
 
+        /**
+         * Must be thread-safe.
+         */
         @Nullable
         protected abstract V maybeGetRetainedValue(K key);
 
         @Nullable
         @Override
         public V getIfPresent(K key) {
-            synchronized (lock) {
-                return getIfPresentWithoutLock(key);
-            }
+            return valuesForThisSession
+                .computeIfAbsent(key, k -> Lazy.unsafe().of(() -> maybeGetRetainedValue(k)))
+                .get();
```

Comment:
~~Calling `get` on an `unsafe` Lazy _outside_ of `computeIfAbsent` does not seem thread-safe~~

I see this is safe because `maybeGetRetainedValue` is safe, but I the mental load of thinking seeing `unsafe` about thread-safety here is unnecessary IMO, and could be solved with a `Lazy.ofValue`